### ISSUE TITLE
plugins implementing their own Provider types

### ISSUE THREAD

**Issue by Lanchon** [gradle/gradle#9184]
hey @big-guy,

you've said [here](https://github.com/gradle/gradle/issues/6244#issuecomment-411402456):
> We do not want plugins to implement their own `Provider` types.

the android plugin was changed from v3.2 to v3.3 to use the lazy task config API. i'm hooking that plugin and i want my code to support both android plugin versions.

my plugin is lazy. to be used with android plugin v3.2 it needs [these adapters](https://github.com/DexPatcher/dexpatcher-gradle/blob/v2/src/main/groovy/lanchon/dexpatcher/gradle/VariantHelper.groovy#L56-L88), which in turn use [my own `Provider` implementation](https://github.com/DexPatcher/dexpatcher-gradle/blob/e0174c0dbac1c0670e9290f71ecf176669852ebd/src/main/groovy/lanchon/dexpatcher/gradle/ExistingTaskProvider.groovy).

what's wrong with all this and how can i fix it?

**Comment by lacasseio** [comment]
You could ask the `TaskContainer` for the `TaskProvider` instead with something like `variant.packageApplication.project.tasks.named(variant.packageApplication.name)` inside your adapters.

**Comment by Lanchon** [comment]
depending on task names is brittle. i won't do it, there is an API for it so i'll use it.

you have to admit that people need to interact with code bases outside of what gradle provides. and in doing so they might need to interact with legacy code not converted to the lazy API. and then, implementing your own provider is necessary or at least the cleanest solution by far.

you are suggesting an inferior solution (task names) to what i have so i'll keep mine. a better solution on your part would be to extend Task with a getProvider(), because there **is** a use case for it, instead of so quickly dismissing this issue. but suit yourself, it's your product. on my part i'll make sure not to create any more issues to help you because it seems unwanted. i'll only create issues when i need help.

**Comment by Lanchon** [comment]
`task.project.tasks.named(task.name)` works but is kinda funky. a better solution on your part would be to extend Task with a getProvider() because there is a use case for it.

**Comment by Lanchon** [comment]
i've switched:

https://github.com/DexPatcher/dexpatcher-gradle/compare/2dd2ae2132383e8461e08e303c7173d534b86eb1...c7119596d855b89af00fe0fd7811bc04e31dd8ed

but i'd still like to avoid a wasteful lookup by name. consider `Task.getProvider()` or similar.
